### PR TITLE
feat: generate lock sudt tx

### PR DESCRIPTION
**Pull request by pygman**

### CODE REVIEW

File: offchain-modules/packages/x/src/ckb/tx-helper/generator.ts [hunk 1]
```diff
@@ -613,13 +613,145 @@ export class CkbTxGenerator extends CkbTxHelper {
   }
 
   async lockSUDT(
```
Comment (mkxbl) [inline offchain-modules/packages/x/src/ckb/tx-helper/generator.ts:615]: 1. Forgot to handle user sudt change cell? If user want to lock 100 sudt, and inputs collect 120 sudt of user, then outputs should refund 20 sudt to user sudt change cell
2. Should calculate and handle tx fee

File: offchain-modules/packages/x/src/ckb/tx-helper/generator.ts [hunk 2]
```diff
@@ -613,13 +613,184 @@ export class CkbTxGenerator extends CkbTxHelper {
...
+      };
+      senderOutputSudtChangeCellMinimalCapacity = minimalCellCapacity(senderOutputSudtChangeCell);
+      senderOutputSudtChangeCell.cell_output.capacity = `0x${senderOutputSudtChangeCellMinimalCapacity.toString(16)}`;
+      // add sudt change output
+      txSkeleton = txSkeleton.update('outputs', (outputs) => {
+        return outputs.push(senderOutputSudtChangeCell);
+      });
+    }
+    // collect ckb
+    const senderNeedCkbSupplyCapacity =
```
Comment (mkxbl) [inline offchain-modules/packages/x/src/ckb/tx-helper/generator.ts:730]: Calculate logic is wrong for `senderNeedCkbSupplyCapacity`. 
```
const outputNeedCapacity = bridgeFee + senderOutputSudtChangeCellMinimalCapacity + senderOutputCkbChangeCellMinimalCapacity + 1n * BigInt(100000000)`;

if (totalCapacityOfSudtCellsInInputs >= outputNeedCapacity) {
  senderNeedCkbSupplyCapacity = 0; 
  // next: refund change capacity to senderOutputCkbChangeCell
} else {
  senderNeedCkbSupplyCapacity = outputNeedCapacity - totalCapacityOfSudtCellsInInputs;
  // next: collect capacity cells of user, then refund change capacity to senderOutputCkbChangeCell
}


```

Comment (pygman) [reply offchain-modules/packages/x/src/ckb/tx-helper/generator.ts]: fixed this logic, collect ckb when inputs capacity does not offset outputs capacity, and to refund directly when the input capacity is sufficient

File: offchain-modules/packages/x/src/ckb/tx-helper/generator.ts (hunk 1 above)
Comment (pygman) [reply offchain-modules/packages/x/src/ckb/tx-helper/generator.ts:615]: 1. sudt change with minimal cell capacity is written in the second output (if needed), ckb change with removal fee and sudt change cell capacity are written in the last output
2. add the calculation of the tx fee at the end

File: offchain-modules/packages/x/src/ckb/tx-helper/generator.ts [hunk 3]
```diff
@@ -613,13 +613,145 @@ export class CkbTxGenerator extends CkbTxHelper {
...
+      );
+    txSkeleton = txSkeleton.update('inputs', (inputs) => {
+      return inputs.push(...searchedSenderFeeCells);
+    });
+
+    // add outputs
+    const committeeMultisigLockscript = parseAddress(getOmniLockMultisigAddress());
+    const committeeMultisigCell: Cell = {
+      cell_output: {
+        capacity: `0x${(amount + bridgeFee).toString(16)}`,
```
[the remark this answers is not in the document]
Comment (pygman) [reply offchain-modules/packages/x/src/ckb/tx-helper/generator.ts]: sudt amount moved into data,capacity of committee multisig cell using bridge fee

File: offchain-modules/packages/x/src/ckb/tx-helper/generator.ts [hunk 4]
```diff
@@ -613,13 +613,146 @@ export class CkbTxGenerator extends CkbTxHelper {
...
+    const sudtCells = await this.collector.collectSudtByAmount(searchKey, amount);
+    const total = sudtCells.map((cell) => utils.readBigUInt128LE(cell.data)).reduce((a, b) => a + b, 0n);
+    if (total < amount) {
+      throw new Error('sudt amount is not enough!');
+    }
+    logger.debug('lock sudtCells: ', sudtCells);
+    txSkeleton = txSkeleton.update('inputs', (inputs) => {
+      return inputs.concat(sudtCells);
+    });
+    // add inputs
```
[the remark this answers is not in the document]
Comment (pygman) [reply offchain-modules/packages/x/src/ckb/tx-helper/generator.ts]: modified txSkeleton update order and comment position

File: offchain-modules/packages/x/src/ckb/tx-helper/generator.ts [hunk 5]
```diff
@@ -613,13 +613,203 @@ export class CkbTxGenerator extends CkbTxHelper {
...
+      };
+      senderOutputSudtChangeCellMinimalCapacity = minimalCellCapacity(senderOutputSudtChangeCell);
+      senderOutputSudtChangeCell.cell_output.capacity = `0x${senderOutputSudtChangeCellMinimalCapacity.toString(16)}`;
+      // add sudt change output
+      txSkeleton = txSkeleton.update('outputs', (outputs) => {
+        return outputs.push(senderOutputSudtChangeCell);
+      });
+    }
+    // collect ckb
+    const totalCapacityOfSudtCellsInInputs = sudtCells
```
[the remark this answers is not in the document]
Comment (pygman) [reply offchain-modules/packages/x/src/ckb/tx-helper/generator.ts:735]: fixed, calculate cell capacity by cell_output.capacity

**Review by mkxbl — APPROVED** [review]

**Pull request merged**